File: src/test/java/neoflix/_08_FavoriteFlagTest.java
// Task: Rewrite the AuthService to allow users to authenticate against the database
// Outcome: A user will be able to authenticate against their database record
package neoflix;

import neoflix.services.FavoriteService;
import neoflix.services.MovieService;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.neo4j.driver.Driver;
import org.neo4j.driver.Values;

import static neoflix.Params.Order.DESC;
import static neoflix.Params.Sort.imdbRating;
import static org.junit.jupiter.api.Assertions.*;

class _08_FavoriteFlagTest {
    private static Driver driver;

    private static final String userId = "fe770c6b-4034-4e07-8e40-2f39e7a6722c";
    private static final String email = "graphacademy.flag@neo4j.com";

    @BeforeAll
    static void initDriver() {
        AppUtils.loadProperties();
        driver = AppUtils.initDriver();

        driver.session().writeTransaction(tx -> tx.run("""
                MERGE (u:User {userId: $userId}) SET u.email = $email
                """, Values.parameters("userId", userId, "email", email)));
    }

    @AfterAll
    static void closeDriver() {
        driver.close();
    }

    @BeforeEach
    void setUp() {
        try (var session = driver.session()) {
            session.writeTransaction(tx ->
                    tx.run("MATCH (u:User {userId: $userId})-[r:HAS_FAVORITE]->(m:Movie) DELETE r",
                            Values.parameters("userId", userId)));
        }
    }

    @Test
    void favoriteMovieReturnsFlaggedInMovieList() {
        MovieService movieService = new MovieService(driver);
        FavoriteService favoriteService = new FavoriteService(driver);

        // Get the most popular movie
        var topMovie = movieService.all(new Params(null, imdbRating, DESC, 1, 0), userId);

        // Add top movie to user favorites
        var topMovieId = topMovie.get(0).get("tmdbId").toString();
        var add = favoriteService.add(userId, topMovieId);
        assertEquals(topMovieId, add.get("tmdbId"));
        assertTrue((Boolean)add.get("favorite"), "top movie is favorite");

        var addCheck = favoriteService.all(userId, new Params(null, imdbRating, Params.Order.ASC, 999, 0));

        assertEquals(1, addCheck.size());
        var found = addCheck.stream().allMatch(movie -> movie.get("tmdbId").equals(topMovieId));
        assertTrue(found);

        var topTwo = movieService.all(new Params(null, imdbRating, DESC, 2, 0), userId);
        assertEquals(topMovieId, topTwo.get(0).get("tmdbId"));
        assertEquals(true, topTwo.get(0).get("favorite"));
        assertEquals(false, topTwo.get(1).get("favorite"));
    }
}

File: pkg/challenges/08_favorite_flag_test.go
package challenges_test

import (
	"github.com/neo4j-graphacademy/neoflix/pkg/fixtures"
	"testing"

	"github.com/neo4j-graphacademy/neoflix/pkg/config"
	"github.com/neo4j-graphacademy/neoflix/pkg/routes/paging"
	"github.com/neo4j-graphacademy/neoflix/pkg/services"
	"github.com/neo4j/neo4j-go-driver/v4/neo4j"
)

func TestFavoritesFlag(t *testing.T) {
	// Load Settings
	settings, err := config.ReadConfig("../../config.json")
	assertNilError(t, err)

	// Init Driver
	driver, err := config.NewDriver(settings)
	assertNilError(t, err)

	defer func() {
		assertNilError(t, driver.Close())
	}()

	fixtureLoader := &fixtures.FixtureLoader{Prefix: "../.."}
	// Create Services
	favoriteService := services.NewFavoriteService(
		fixtureLoader,
		driver)
	movieService := services.NewMovieService(
		fixtureLoader,
		driver)

	assertNotNil(t, favoriteService)
	assertNotNil(t, movieService)

	userId := "fe770c6b-4034-4e07-8e40-2f39e7a6722c"
	email := "graphacademy.flag@neo4j.com"

	// Create the User
	session := driver.NewSession(neo4j.SessionConfig{})
	session.Run(`
		MERGE (u:User {userId: $userId}) SET u.email = $email
		FOREACH (r IN [(u)-[r:HAS_FAVORITE]->() | r ] | DELETE r)
	`, map[string]interface{}{"userId": userId, "email": email})

	// Get the most popular movie
	firstCall, err := movieService.FindAll(userId, paging.NewPaging("", "imdbRating", "DESC", 0, 1))

	assertNilError(t, err)
	assertNotNil(t, firstCall)

	movieId := firstCall[0]["tmdbId"].(string)
	assertEquals(t, false, firstCall[0]["favorite"])

	// Add it to user favorites
	favorite, err := favoriteService.Save(userId, movieId)

	assertNilError(t, err)

	assertEquals(t, movieId, favorite["tmdbId"])
	assertEquals(t, true, favorite["favorite"])

	// Get most popular movie again
	secondCall, err := movieService.FindAll(userId, paging.NewPaging("", "imdbRating", "DESC", 0, 1))

	assertNilError(t, err)
	assertNotNil(t, secondCall)

	assertEquals(t, movieId, secondCall[0]["tmdbId"])
	assertEquals(t, true, secondCall[0]["favorite"])
}
